src/Ocelot/Multiplexer/MultiplexingMiddleware.cs

```diff
@@ -1,219 +1,265 @@
+    /// <returns>True if only the first downstream route should be processed.</returns>
+    private static bool ShouldProcessSingleRoute(HttpContext context, ICollection routes) => context.WebSockets.IsWebSocketRequest || routes.Count == 1;
+
+    /// <summary>
+    /// Processing a single downstream route (no route keys).
+    /// In that case, no need to make copies of the http context.
+    /// </summary>
+    /// <param name="context">The http context.</param>
+    /// <param name="route">The downstream route.</param>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+    protected virtual Task ProcessSingleRoute(HttpContext context, DownstreamRoute route)
+    {
+        context.Items.UpsertDownstreamRoute(route);
+        return _next.Invoke(context);
+    }
 
-                await Task.WhenAll(tasks);
+    /// <summary>
+    /// Processing the downstream routes (no route keys).
+    /// </summary>
+    /// <param name="context">The main http context.</param>
+    /// <param name="route">The route.</param>
+    private async Task ProcessRoutes(HttpContext context, Route route)
+    {
+        var tasks = route.DownstreamRoute.Select(downstreamRoute => ProcessRouteAsync(context, downstreamRoute, _next))
+            .ToArray();
+        var contexts = await Task.WhenAll(tasks);
+        await Map(context, route, [.. contexts]);
+    }
 
-                var contexts = new List<HttpContext>();
+    /// <summary>
+    /// When using route keys, the first route is the main route and the rest are additional routes.
+    /// Since we need to break if the main route response is null, we must process the main route first.
+    /// </summary>
+    /// <param name="context">The http context.</param>
+    /// <param name="route">The first route, the main route.</param>
+    /// <returns>The updated http context.</returns>
+    private Task<HttpContext> ProcessMainRoute(HttpContext context, DownstreamRoute route)
+    {
+        context.Items.UpsertDownstreamRoute(route);
+        return Fire(context, _next);
+    }
 
-                foreach (var task in tasks)
-                {
-                    var finished = await task;
-                    contexts.Add(finished);
-                }
+    /// <summary>
+    /// Processing the downstream routes with route keys except the main route that has already been processed.
+    /// </summary>
+    /// <param name="context">The main http context.</param>
+    /// <param name="routes">The downstream routes.</param>
+    /// <param name="routeKeysConfigs">The route keys config.</param>
+    /// <param name="mainResponse">The response from the main route.</param>
+    /// <returns>A list of the tasks' http contexts.</returns>
+    protected virtual async Task<HttpContext[]> ProcessRoutesWithRouteKeys(HttpContext context,
+        IEnumerable<DownstreamRoute> routes, IReadOnlyCollection<AggregateRouteConfig> routeKeysConfigs,
+        HttpContext mainResponse)
```

Comment:
Not an issue. Just question.
Is it our lovely new feature aka **Complex Aggregation**?

Reply:
It's not a new feature, it's a not documented, existing, feature, that we now can understand because we decided to clean up the code.